### ISSUE TITLE
Decode probrem

### ISSUE THREAD

**Issue by cz3344** [seanie12/nqg#1]
Hi @seanie12 ,
I used your code trying to replicate the Paper's results.
Unfortunately, I got a lot of UNK in the generated file.
Can you kindly share your reference.py code for QG?

> UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN ?
UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN ' ?
UNKNOWN UNKNOWN UNKNOWN UNKNOWN who play UNKNOWN UNKNOWN UNKNOWN UNKNOWN ?
UNKNOWN UNKNOWN is UNKNOWN UNKNOWN UNKNOWN ' UNKNOWN ?
UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN ' ?
UNKNOWN UNKNOWN UNKNOWN is UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN UNKNOWN ' ?
how many miles is UNKNOWN international airport located ?
UNKNOWN UNKNOWN airport is connected to what light system ?
how many UNKNOWN does UNKNOWN international airport have ?
how many UNKNOWN does UNKNOWN international airport have per year ?
how many UNKNOWN are available in UNKNOWN international airport ?
what was the name of UNKNOWN UNKNOWN ?
how many grade UNKNOWN stations in the uk ?
who was the first UNKNOWN UNKNOWN in 1850 ?

**Comment by seanie12** [comment]
Hi. Can you tell me your setting in more detail? What do you mean reference.py?

**Comment by cz3344** [comment]
@seanie12 Sry. It's inference.py. The one in github right now is using BERT related settings(QGTrainer(), BERT tokenizer and pre-trained word embedding). 
I modified some place to fit the Trainer()@trainer.py.
In terms of the setting, I followed your config.py file but set the vocab_size to 45000, embedding_size to 300 and using GLOVE embedding according the paper.

**Comment by cz3344** [comment]
`class BeamSearcher(object):
    def __init__(self, model_path, output_dir):
        with open(config.embedding, "rb") as f:
            embedding = pickle.load(f)
            embedding = torch.Tensor(embedding).to(config.device)
        with open(config.word2idx_file, "rb") as f:
            word2idx = pickle.load(f)
        self.word2idx = word2idx
        self.idx2word = {v: k for k, v in self.word2idx.items()}

        print("load test data")
        self.data_loader = get_loader(config.test_src_file,
                                           config.test_trg_file,
                                           self.word2idx,
                                           use_tag=config.use_tag,
                                           batch_size=1,
                                           debug=config.debug)
        print("load trained model from {}".format(model_path))
        self.model = Seq2seq(config.dropout, embedding, config.use_tag, model_path=model_path)
        self.model.requires_grad = False
        self.model.eval_mode()
        self.pred_file = output_dir + "/generated.txt"
        self.src_file = output_dir + "/src.txt"
        self.golden_file = output_dir + "/gloden_question.txt"
        self.ans_file = output_dir + "answer.txt"
        self.total_file = output_dir + "all_files.csv"
        if not os.path.exists(output_dir):
            os.makedirs(output_dir)

    @staticmethod
    def sort_hypotheses(hypotheses):
        return sorted(hypotheses, key=lambda h: h.avg_log_prob, reverse=True)


    def decode(self):
        pred_fw = open(self.pred_file, "w", encoding="utf-8")
        golden_fw = open(self.golden_file, "w", encoding="utf-8")
        src_fw = open(self.src_file, "w", encoding="utf-8")
        ans_fw = open(self.ans_file, "w", encoding="utf-8")
        for i, eval_data in enumerate(tqdm(self.data_loader)):
            if config.use_tag:
                src_seq, ext_src_seq, src_len, trg_seq, ext_trg_seq, trg_len, tag_seq, _ = eval_data
            else:
                src_seq, ext_src_seq, src_len, trg_seq, ext_trg_seq, trg_len, _ = eval_data
                tag_seq = None
            src_len = torch.LongTensor(src_len)

            if config.use_gpu:
                src_seq = src_seq.to(config.device)
                ext_src_seq = ext_src_seq.to(config.device)
                src_len = src_len.to(config.device)
                trg_seq = trg_seq.to(config.device)
                ext_trg_seq = ext_trg_seq.to(config.device)
                if config.use_tag:
                    tag_seq = tag_seq.to(config.device)
                else:
                    tag_seq = None

            best_question = self.beam_search(src_seq, ext_src_seq, src_len, tag_seq)


            # discard START  token
            output_indices = [int(idx) for idx in best_question.tokens[1:]] # drop sos
            decoded_words = [self.idx2word[i] for i in output_indices]
            try:
                fst_stop_idx = decoded_words.index("EOS")
                decoded_words = decoded_words[:fst_stop_idx]
            except ValueError:
                decoded_words = decoded_words
            decoded_words = " ".join(decoded_words)

            q_id = trg_seq[1:-1].cpu().numpy()
            golden_question = [self.idx2word[i] for i in q_id]
            golden_question = " ".join(golden_question)

            pred_fw.write(decoded_words + "\n")
            golden_fw.write(" ".join(golden_question) + "\n")

        pred_fw.close()
        golden_fw.close()
        # self.merge_files(self.total_file)

    def beam_search(self, src_seq, ext_src_seq, src_len, tag_seq):
        # forward encoder
        enc_outputs, enc_states = self.model.encoder(src_seq, src_len, tag_seq)
        h, c = enc_states  # [2, b, d] but b = 1; [2, 1, 600]
        hypotheses = [Hypothesis(tokens=[self.word2idx["<s>"]],
                                 log_probs=[0.0],
                                 state=(h[:, 0, :], c[:, 0, :]),
                                 context=None) for _ in range(config.beam_size)] # 改
        # tile enc_outputs, enc_mask for beam search
        ext_src_seq = ext_src_seq.repeat(config.beam_size, 1)
        enc_outputs = enc_outputs.repeat(config.beam_size, 1, 1)
        zeros = enc_outputs.sum(dim=-1)
        enc_mask = (zeros == 0).byte()
        enc_features = self.model.decoder.get_encoder_features(enc_outputs)

        num_steps = 0
        results = []
        while num_steps < config.max_decode_step and len(results) < config.beam_size:
            latest_tokens = [h.latest_token for h in hypotheses]
            prev_y = torch.LongTensor(latest_tokens).view(-1)

            if config.use_gpu:
                prev_y = prev_y.to(config.device)

            # make batch of which size is beam size
            all_state_h = []
            all_state_c = []
            for h in hypotheses:
                state_h, state_c = h.state  # [num_layers, d]
                all_state_h.append(state_h)
                all_state_c.append(state_c)

            prev_h = torch.stack(all_state_h, dim=1)  # [num_layers, beam, d]
            prev_c = torch.stack(all_state_c, dim=1)  # [num_layers, beam, d]
            prev_states = (prev_h, prev_c)
            # [beam_size, |V|]
            logits, states, = self.model.decoder.decode(prev_y,
                                                        ext_src_seq,
                                                        prev_states,
                                                        enc_features,
                                                        enc_mask)
            h_state, c_state = states
            log_probs = F.log_softmax(logits, dim=1)
            top_k_log_probs, top_k_ids \
                = torch.topk(log_probs, config.beam_size * 2, dim=-1)

            all_hypotheses = []
            num_orig_hypotheses = 1 if num_steps == 0 else len(hypotheses)
            for i in range(num_orig_hypotheses):
                h = hypotheses[i]
                state_i = (h_state[:, i, :], c_state[:, i, :])
                for j in range(config.beam_size * 2):
                    new_h = h.extend(token=top_k_ids[i][j].item(),
                                     log_prob=top_k_log_probs[i][j].item(),
                                     state=state_i,
                                     context=None)
                    all_hypotheses.append(new_h)

            hypotheses = []
            for h in self.sort_hypotheses(all_hypotheses):
                if h.latest_token == self.word2idx["EOS"]:
                    if num_steps >= config.min_decode_step:
                        results.append(h)
                else:
                    hypotheses.append(h)

                if len(hypotheses) == config.beam_size or len(results) == config.beam_size:
                    break
            num_steps += 1
        if len(results) == 0:
            results = hypotheses
        h_sorted = self.sort_hypotheses(results)

        return h_sorted[0]`

**Comment by seanie12** [comment]
I updated inference.py and trainer.py. So i hope you can reproduce the result

**Comment by cz3344** [comment]
@seanie12  Thx for your help, Seanie12. I reproduced the results.
